File: src/learn-topic-wise/08-trees/utils/generate-trees.cpp
#include "generate-trees.h"

#include <iostream>
#include <optional>
#include <queue>
#include <stdexcept>
#include <vector>

using namespace std;

namespace utils {

// Define constructors
TreeNode::TreeNode() : val(0), left(nullptr), right(nullptr) {}
TreeNode::TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
TreeNode::TreeNode(int x, TreeNode* left, TreeNode* right) : val(x), left(left), right(right) {}

TreeNode* TreeOperations::generate(vector<optional<int>> inArr) {
    int n = inArr.size();

    if (n == 0 || !inArr[0].has_value()) {
        throw invalid_argument("Input must contain a non-null root value.");
    }

    TreeNode* root = new TreeNode(inArr[0].value());

    queue<TreeNode*> q;
    q.push(root);

    int i = 1;
    while (q.size() > 0 && i < n) {
        TreeNode* cur = q.front();
        q.pop();

        if (i < n && inArr[i].has_value()) {
            auto l = new TreeNode(inArr[i].value());
            cur->left = l;
            q.push(l);
        }
        ++i;

        if (i < n && inArr[i].has_value()) {
            auto r = new TreeNode(inArr[i].value());
            cur->right = r;
            q.push(r);
        }
        ++i;
    }

    return root;
};

void TreeOperations::print(TreeNode* root) {
    queue<TreeNode*> q;
    q.push(root);

    while (q.size()) {
        int qLen = q.size();

        cout << endl;

        for (int i = 0; i < qLen; ++i) {
            TreeNode* cur = q.front();
            q.pop();

            if (cur == nullptr) continue;

            cout << cur->val << " ";

            if (cur->left != nullptr) {
                q.push(cur->left);
            }
            if (cur->right != nullptr) {
                q.push(cur->right);
            }
        }
    }
}

}  // namespace utils

File: src/learn-topic-wise/08-trees/utils/generate_trees.py
from typing import List


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Utils:
    class TreeOperations:
        def generate(self, arr: List[int]) -> TreeNode:
            if arr[0] == None:
                return None

            n = len(arr)

            root = TreeNode(arr[0])

            i = 1
            q = [root]
            while len(q) > 0 and i < n:
                cur = q.pop(0)

                if arr[i] != None:
                    cur.left = TreeNode(arr[i])
                    q.append(cur.left)
                i += 1

                if i < n and arr[i] != None:
                    cur.right = TreeNode(arr[i])
                    q.append(cur.right)
                i += 1

            return root

        def find(self, root: TreeNode, val: int) -> TreeNode:

            return root

        def print(self, root: TreeNode) -> None:
            if root == None:
                return

            print(root.val)
